clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -476,14 +479,22 @@ public <T> List<T> getConfiguredInstances(List<String> classNames, Class<T> t, M
             return objects;
         Map<String, Object> configPairs = originals();
         configPairs.putAll(configOverrides);
-        for (Object klass : classNames) {
-            Object o = getConfiguredInstance(klass, t, configPairs);
-            objects.add(t.cast(o));
+
+        try {
+            for (Object klass : classNames) {
+                Object o = getConfiguredInstance(klass, t, configPairs);
+                objects.add(t.cast(o));
+            }
+        } catch (Exception e) {
+            for (Object object : objects) {
+                Utils.closeQuietly((AutoCloseable) object, "AutoCloseable object constructed and configured during failed call to getConfiguredInstances");
```

Comment:
We still need an `instanceof` check here:
```suggestion
                if (object instanceof AutoCloseable) {
                    Utils.closeQuietly((AutoCloseable) object, "AutoCloseable object constructed and configured during failed call to getConfiguredInstances");
                }
```

Reply:
Actually, come to think of it, we might also want to invoke `close` on objects created in `getConfiguredInstance` if they throw an exception from `configure` [here](https://github.com/apache/kafka/blob/6c98544a964b40ede6bbe1b3440f8e5db96a4ad6/clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java#L407).

Perhaps we could pull this out into a reusable method and use it both here and there? Thinking something like:

```java
private static void maybeClose(Object object, String name) {
    if (object instanceof AutoCloseable)
        Utils.closeQuietly(object, name);
}
```